### PR TITLE
Added an "unsafe" way to retrieve a byte array from `SdkBytes` and `ResponseBytes` without copying the data.

### LINKED ISSUE #1959: Allow to get array from BytesWrapper without copying
In performance sensitive applications in is important to reduce number of data-copies that happen during processing.
Unfortunately currently when we get data from S3 and then try to use response byte array we have to:
* Either do an array copy
* Get a readonly ByteBuffer - which doesn't provide access to array and forces us to do an array copy later

## Describe the Feature
Provide a method in `BytesWrapper` that allows users to get either underlying array directly or a non-read-only `ByteBuffer`

## Is your Feature Request related to a problem?
Current S3 client implementation does multiple data copies - this is just one more of them, and this reduces performace.

## Describe alternatives you've considered
Unfortunately there are not many alternatives: in some cases one needs access to arrays, for example if one tries top use JNI compression implementations.

## Additional Context

- [x] I may be able to implement this feature request

## Your Environment

* AWS Java SDK version used:
* JDK version used:
* Operating System and version:

### PR DESCRIPTION
**Pull request by millems**
Reduced the number of payload data copies on some code paths.

Fixes #1959.

### CODE REVIEW

**Comment by millems** [comment]
Integration tests have passed. Running benchmarks...

**Comment by millems** [comment]
Unit tests have passed. Still waiting on benchmarks...

**Comment by millems** [comment]
No appreciable difference in benchmarks.

File: core/sdk-core/src/test/java/software/amazon/awssdk/core/ResponseBytesTest.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package software.amazon.awssdk.core;
```
Comment (zoewangg) [inline core/sdk-core/src/test/java/software/amazon/awssdk/core/ResponseBytesTest.java:1]: missing copyright header?

File: .changes/next-release/feature-AWSSDKforJavav2-47fe81e.json [hunk 2]
```diff
@@ -0,0 +1,5 @@
+{
+    "type": "feature",
+    "category": "AWS SDK for Java v2",
+    "description": "Slight performance improvement, from reduced the number of data copies on some code paths."
```
[the remark this answers is not in the document]
Comment (millems) [reply .changes/next-release/feature-AWSSDKforJavav2-47fe81e.json]: Done, as well as removed the actual changes that use the new methods.

File: core/sdk-core/src/test/java/software/amazon/awssdk/core/ResponseBytesTest.java (hunk 1 above)
Comment (millems) [reply core/sdk-core/src/test/java/software/amazon/awssdk/core/ResponseBytesTest.java:16]: Fixed, thanks!

**Pull request merged**